File: src/shared/radius-broadcaster.ts
import { debugLog } from './debug-log';
import { sendValkeyArrayCommand } from './valkey-client.js';
import { wsSender, type WsSendResult } from './ws-sender.js';

const ZONE_ID = 'zone01';
const DEFAULT_LIMIT = 500;

// ---- key builders --------------------------------------------------------

function buildGeoClientsKey(appId: string): string {
  return `msight:${ZONE_ID}:${appId}:geo:clients`;
}

function buildClientKey(appId: string, clientId: string): string {
  return `msight:${ZONE_ID}:${appId}:client:${clientId}`;
}

// ---- types ---------------------------------------------------------------

export type WsInfo = {
  clientId: string;
  connectionId: string;
  domainName: string;
  stage: string;
};

export type RadiusBroadcastParams = {
  appId: string;
  eventId: string;
  message: unknown;
  lat: number;
  lon: number;
  radiusM: number;
  limit?: number;
};

export type RadiusBroadcastResult = {
  nearbyClientCount: number;
  websocketCandidateCount: number;
  deliveredCount: number;
  failedCount: number;
};

// ---- RadiusBroadcaster ---------------------------------------------------

// Find nearby clients and resolve their WebSocket connections in ONE round trip.
//
// This used to be a GEOSEARCH followed by one HMGET per client returned. The
// HMGETs were issued through Promise.all, which reads as though they run in
// parallel — but valkey-client.ts keeps a single command in flight and queues
// the rest, so they were strictly sequential. Fifty nearby clients meant
// fifty-one serial round trips, on a path invoked once per SPaT message per
// app, at roughly 10 Hz per sensor.
//
// Both lookups belong to one logical read anyway, so doing them server-side
// also makes the result self-consistent: no client can connect or drop between
// the geo search and the field fetch.
//
// The derived client keys are not declared in KEYS[]. That is safe on the
// single-node replication group this deploys against, but it is the one thing
// that would have to change first if the cache ever moved to cluster mode,
// where every key a script touches must hash to the same slot.
const NEARBY_WS_CLIENTS_LUA = `
local ids = redis.call('GEOSEARCH', KEYS[1],
  'FROMLONLAT', ARGV[1], ARGV[2],
  'BYRADIUS', ARGV[3], 'm',
  'COUNT', ARGV[4], 'ASC')

local connected = {}
for i = 1, #ids do
  local fields = redis.call('HMGET', ARGV[5] .. ids[i],
    'ws_connection_id', 'ws_domain_name', 'ws_stage')
  if fields[1] and fields[2] and fields[3] then
    connected[#connected + 1] = ids[i]
    connected[#connected + 1] = fields[1]
    connected[#connected + 1] = fields[2]
    connected[#connected + 1] = fields[3]
  end
end

-- Returns the total found alongside the connected subset, so the caller can
-- still report how many clients were nearby but had no live socket.
return { #ids, connected }
`;

export class RadiusBroadcaster {
  private async findNearbyWsClients(
    appId: string,
    lat: number,
    lon: number,
    radiusM: number,
    limit: number
  ): Promise<{ nearbyClientCount: number; wsInfos: WsInfo[] }> {
    const response = await sendValkeyArrayCommand([
      'EVAL',
      NEARBY_WS_CLIENTS_LUA,
      1,
      buildGeoClientsKey(appId),
      lon,
      lat,
      radiusM,
      limit,
      `msight:${ZONE_ID}:${appId}:client:`,
    ]);

    if (!Array.isArray(response) || response.length < 2) {
      return { nearbyClientCount: 0, wsInfos: [] };
    }

    const nearbyClientCount = Number(response[0]) || 0;
    const flat = Array.isArray(response[1]) ? response[1] : [];

    const wsInfos: WsInfo[] = [];
    for (let index = 0; index + 3 < flat.length; index += 4) {
      const [clientId, connectionId, domainName, stage] = flat.slice(index, index + 4);
      if (
        typeof clientId === 'string' &&
        typeof connectionId === 'string' &&
        typeof domainName === 'string' &&
        typeof stage === 'string'
      ) {
        wsInfos.push({ clientId, connectionId, domainName, stage });
      }
    }

    return { nearbyClientCount, wsInfos };
  }

  private async clearWsInfoIfSameConnection(
    appId: string,
    clientId: string,
    connectionId: string
  ): Promise<void> {
    const clientKey = buildClientKey(appId, clientId);
    // Atomic check-and-delete via Lua script — prevents a reconnect that writes
    // a new ws_connection_id between HGET and HDEL from being silently wiped.
    await sendValkeyArrayCommand([
      'EVAL',
      `local cur = redis.call('HGET', KEYS[1], 'ws_connection_id')
 if cur == ARGV[1] then
   redis.call('HDEL', KEYS[1], 'ws_connection_id', 'ws_domain_name', 'ws_stage', 'ws_connected_at', 'ws_status')
   return 1
 end
 return 0`,
      1,
      clientKey,
      connectionId,
    ]);
  }

  async broadcast(params: RadiusBroadcastParams): Promise<RadiusBroadcastResult> {
    const { appId, eventId, message, lat, lon, radiusM, limit = DEFAULT_LIMIT } = params;

    debugLog('radius broadcast start', {
      appId,
      eventId,
      lat,
      lon,
      radiusM,
      limit,
      geoClientsKey: buildGeoClientsKey(appId),
    });

    // 1. Find nearby clients and their live sockets — a single round trip.
    const { nearbyClientCount, wsInfos } = await this.findNearbyWsClients(
      appId,
      lat,
      lon,
      radiusM,
      limit
    );

    debugLog('radius broadcast lookup result', {
      appId,
      nearbyClientCount,
      candidateCount: wsInfos.length,
      skippedCount: nearbyClientCount - wsInfos.length,
      candidates: wsInfos.map((info) => ({
        clientId: info.clientId,
        connectionId: info.connectionId,
        domainName: info.domainName,
        stage: info.stage,
      })),
    });

    // 2. Send to all connections in parallel via wsSender
    const sendResult: WsSendResult = await wsSender.send(appId, eventId, message, wsInfos);

    debugLog('radius broadcast send result', {
      appId,
      deliveredCount: sendResult.deliveredCount,
      failedCount: sendResult.failedCount,
      goneConnectionIds: sendResult.goneConnectionIds,
    });

    // 3. Clean up stale Valkey entries for gone connections
    if (sendResult.goneConnectionIds.length > 0) {
      const wsInfoByConnectionId = new Map(wsInfos.map((info) => [info.connectionId, info]));

      await Promise.all(
        sendResult.goneConnectionIds.map(async (connectionId) => {
          const info = wsInfoByConnectionId.get(connectionId);
          if (info) {
            await this.clearWsInfoIfSameConnection(appId, info.clientId, connectionId);
          }
        })
      );
    }

    return {
      nearbyClientCount,
      websocketCandidateCount: wsInfos.length,
      deliveredCount: sendResult.deliveredCount,
      failedCount: sendResult.failedCount,
    };
  }
}

// Module-level singleton — reused across warm Lambda invocations.
export const radiusBroadcaster = new RadiusBroadcaster();

File: src/services/sensor-consumer/radius_broadcaster.py
"""
radius_broadcaster.py

Python equivalent of src/shared/radius-broadcaster.ts + ws-sender.ts.

Provides two public functions:
  - get_nearby_clients(redis_client, app_id, lat, lon, radius_m, limit)
      -> list[WsClient]   (each client has ws connection info resolved)

  - broadcast_to_clients(app_id, clients, message, aws_region, event_id=None)
      -> BroadcastResult

Keeping both operations separate lets the caller do an early-exit before
any state-mutating work (stale-check write) when there are no nearby clients.
"""

import asyncio
import json
import uuid
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional

import boto3
from botocore.exceptions import ClientError

# ---- constants (mirror radius-broadcaster.ts) ----------------------------

_ZONE_ID       = "zone01"
_DEFAULT_LIMIT = 500
MAX_WORKERS     = 20  # for ThreadPoolExecutor in broadcast_to_clients()

# Module-level boto3 client cache — HTTP connections are kept alive and reused
# across consecutive broadcast cycles, matching TypeScript WsSender behaviour.
# endpoint (str) -> boto3 management api client
_apigw_clients: dict = {}

# ---- key builders --------------------------------------------------------

def _geo_clients_key(app_id: str) -> str:
    return f"msight:{_ZONE_ID}:{app_id}:geo:clients"


def _client_key(app_id: str, client_id: str) -> str:
    return f"msight:{_ZONE_ID}:{app_id}:client:{client_id}"


# ---- data types ----------------------------------------------------------

@dataclass
class WsClient:
    client_id:     str
    connection_id: str
    domain_name:   str
    stage:         str


@dataclass
class BroadcastResult:
    delivered_count: int = 0
    failed_count:    int = 0
    gone_count:      int = 0


# ---- public: find nearby clients -----------------------------------------

def get_nearby_clients(
    redis_client,
    app_id:   str,
    lat:      float,
    lon:      float,
    radius_m: float,
    limit:    int = _DEFAULT_LIMIT,
) -> list[WsClient]:
    """
    Returns WsClient records for every client within radius_m metres of
    (lat, lon) that has an active WebSocket connection registered in Valkey.

    Uses a single GEOSEARCH + pipelined HMGET — no per-client round trips.
    """
    geo_key = _geo_clients_key(app_id)

    nearby_ids: list[str] = redis_client.execute_command(
        "GEOSEARCH", geo_key,
        "FROMLONLAT", lon, lat,
        "BYRADIUS", radius_m, "m",
        "COUNT", limit,
        "ASC",
    )

    if not nearby_ids:
        return []

    # Batch-fetch WS connection info for all nearby clients in one pipeline.
    pipe = redis_client.pipeline(transaction=False)
    for client_id in nearby_ids:
        pipe.hmget(
            _client_key(app_id, client_id),
            "ws_connection_id", "ws_domain_name", "ws_stage",
        )
    field_batches: list[list] = pipe.execute()

    clients: list[WsClient] = []
    for client_id, fields in zip(nearby_ids, field_batches):
        if not fields or len(fields) < 3:
            continue
        connection_id, domain_name, stage = fields[0], fields[1], fields[2]
        if not connection_id or not domain_name or not stage:
            continue  # client has no active WS session
        clients.append(WsClient(
            client_id=client_id,
            connection_id=connection_id,
            domain_name=domain_name,
            stage=stage,
        ))

    return clients


# ---- internal: clear stale WS entry in Valkey ----------------------------

# Lua script: atomically check-and-delete so a reconnect that writes a new
# ws_connection_id between the read and the HDEL cannot be silently wiped.
_CLEAR_IF_SAME_SCRIPT = """
local current = redis.call('HGET', KEYS[1], 'ws_connection_id')
if current == ARGV[1] then
    redis.call('HDEL', KEYS[1],
        'ws_connection_id', 'ws_domain_name', 'ws_stage',
        'ws_connected_at', 'ws_status')
    return 1
end
return 0
"""

def _clear_ws_info_if_same_connection(
    redis_client,
    app_id:        str,
    client_id:     str,
    connection_id: str,
) -> None:
    """Atomically remove WS fields only if the stored connection_id still matches."""
    key = _client_key(app_id, client_id)
    redis_client.eval(_CLEAR_IF_SAME_SCRIPT, 1, key, connection_id)


# ---- internal: send to one client ----------------------------------------

def _send_one(
    apigw_clients: dict,            # endpoint -> boto3 client (shared, pre-built)
    item:          WsClient,
    payload_bytes: bytes,
    aws_region:    str,
) -> tuple[str, bool, bool]:
    """
    Returns (connection_id, delivered: bool, gone: bool).
    Caller is responsible for Valkey cleanup on gone=True.
    """
    endpoint = f"https://{item.domain_name}/{item.stage}"
    if endpoint not in apigw_clients:
        apigw_clients[endpoint] = boto3.client(
            "apigatewaymanagementapi",
            endpoint_url=endpoint,
            region_name=aws_region,
        )
    client = apigw_clients[endpoint]
    try:
        client.post_to_connection(
            ConnectionId=item.connection_id,
            Data=payload_bytes,
        )
        return (item.connection_id, True, False)
    except ClientError as exc:
        code = exc.response.get("Error", {}).get("Code", "")
        if code == "GoneException":
            return (item.connection_id, False, True)
        return (item.connection_id, False, False)
    except Exception:
        return (item.connection_id, False, False)


# ---- public: broadcast to pre-fetched client list ------------------------

async def broadcast_to_clients(
    redis_client,
    app_id:     str,
    clients:    list[WsClient],
    message:    dict,
    aws_region: str,
    event_id:   Optional[str] = None,
) -> BroadcastResult:
    """
    Push *message* to every client in *clients* via API Gateway Management API.
    Gone connections are cleaned up in Valkey automatically.

    *clients* should come from get_nearby_clients() — no GEOSEARCH is repeated.
    """
    if not clients:
        return BroadcastResult()

    if event_id is None:
        event_id = str(uuid.uuid4())

    payload_bytes = json.dumps({
        "app_id":           app_id,
        "event_id":         event_id,
        "message":          message,
        "server_timestamp": datetime.now(timezone.utc).isoformat(),
    }).encode("utf-8")

    result = BroadcastResult()
    gone_items: list[WsClient] = []
    by_connection_id = {c.connection_id: c for c in clients}

    loop = asyncio.get_running_loop()
    # One thread per client so no client's timeout can queue another.
    # Pass the module-level cache so boto3 connections are reused across calls.
    with ThreadPoolExecutor(max_workers=min(len(clients), MAX_WORKERS)) as executor:
        send_results = await asyncio.gather(*[
            loop.run_in_executor(executor, _send_one, _apigw_clients, item, payload_bytes, aws_region)
            for item in clients
        ])

    for _conn_id, delivered, gone in send_results:
        if delivered:
            result.delivered_count += 1
        elif gone:
            result.gone_count += 1
            gone_items.append(by_connection_id[_conn_id])
        else:
            result.failed_count += 1

    # Clean up stale Valkey entries for gone connections.
    for item in gone_items:
        try:
            _clear_ws_info_if_same_connection(redis_client, app_id, item.client_id, item.connection_id)
        except Exception:
            pass

    return result
